pkg/testing/matrix/matrix.go

```diff
@@ -0,0 +1,457 @@
+			if !filepath.IsAbs(provider.Path) {
+				proj.Plugins.Providers[i].Path = filepath.Join(dir, provider.Path)
+			}
+		}
+		for i, language := range proj.Plugins.Languages {
+			if !filepath.IsAbs(language.Path) {
+				proj.Plugins.Languages[i].Path = filepath.Join(dir, language.Path)
+			}
+		}
+
+		for i, analyzer := range proj.Plugins.Analyzers {
+			if !filepath.IsAbs(analyzer.Path) {
+				proj.Plugins.Analyzers[i].Path = filepath.Join(dir, analyzer.Path)
+			}
+		}
+	}
+
+	//Generate SDKs
+	pluginctx, err := plugin.NewContextWithRoot(cmdutil.Diag(), cmdutil.Diag(), nil, pwd, projinfo.Root,
+		projinfo.Proj.Runtime.Options(), false, nil, proj.Plugins)
+	assert.NoError(t, err)
+	assert.NotNil(t, pluginctx)
+	for _, plugin := range opts.Plugins {
+		assert.NotNil(t, plugin.Version)
+
+		if plugin.Kind != workspace.ResourcePlugin {
+			continue
+		}
+
+		info, err := pluginctx.Host.ResolvePlugin(plugin.Kind, plugin.Name, &plugin.Version)
+		assert.NoError(t, err)
+		assert.NotNil(t, info)
+
+		provider, err := pluginctx.Host.Provider(tokens.Package(info.Name), &plugin.Version)
+		assert.NoError(t, err)
+		assert.NotNil(t, provider)
+
+		schemaBytes, err := provider.GetSchema(int(plugin.Version.Major))
+		assert.NoError(t, err)
+		assert.NotNil(t, schemaBytes)
+
+		var spec schema.PackageSpec
+		err = json.Unmarshal(schemaBytes, &spec)
+		assert.NoError(t, err)
+
+		pkg, diags, err := schema.BindSpec(spec, nil)
+		assert.NoError(t, err)
+		assert.Empty(t, diags)
+		assert.NotNil(t, pkg)
+
+		//Kludge
+		/*for _, p := range opts.Plugins {
+			if p.Name == pkg.Name {
+				pkg.Version = &p.Version
+			}
+		}*/
+
+		pkg.Test = true
+
+		pkgName := pkg.Name
+
+		for _, langOpt := range opts.Languages {
+			lang := langOpt.Language
+
+			var files map[string][]byte
+			var err error
+			switch lang {
+			case GO:
+				files, err = gogen.GeneratePackage(pkgName, pkg)
+				assert.NoError(t, err)
+			case PYTHON:
+				err = PythonConfigurePkg(pkg)
+				assert.NoError(t, err)
+				files, err = pygen.GeneratePackage(pkgName, pkg, files)
+				assert.NoError(t, err)
+			case NODEJS:
+				err = NodeConfigurePkg(pkg)
+				assert.NoError(t, err)
+				files, err = jsgen.GeneratePackage(pkgName, pkg, files)
+				assert.NoError(t, err)
+			case DOTNET:
+				err = DotnetConfigurePkg(pkg)
+				assert.NoError(t, err)
+				files, err = dotnetgen.GeneratePackage(pkgName, pkg, files)
+				assert.NoError(t, err)
+			//In the future we should support java but I hava no idea where to even start with that.
+			default:
+				fmt.Printf("Unknown language: '%s'", lang)
+				continue
+			}
+
+			sdkDir := filepath.Join(dir, fmt.Sprintf("%s-sdk", pkgName), lang)
+			for p, file := range files {
+				// TODO: full conversion from path to filepath
+				err = os.MkdirAll(filepath.Join(sdkDir, path.Dir(p)), 0700)
+				assert.NoError(t, err)
+				err = os.WriteFile(filepath.Join(sdkDir, p), file, 0600)
+				assert.NoError(t, err)
+			}
+			if lang == "go" {
```

Comment:
Why are these needed? This feels like something GeneratePackage ought to do? Do we need a new "generateOnly" flag to GeneratePackage so that this logic can live in the language code not here?

Reply:
I added this at some point but I don't remember why! Seems to work okay if I remove it.